src/Printers/Communication/UsbPrinterDeviceChannel.ts

```diff
@@ -0,0 +1,203 @@
+import { WebZlpError } from '../../WebZlpError';
+import { LineBreakTransformer } from './LineBreakTransformer';
+import {
+    PrinterChannelType,
+    IPrinterDeviceChannel,
+    PrinterCommMode,
+    PrinterCommunicationError
+} from './PrinterCommunication';
+
+/** Class for managing the WebUSB communication with a printer. */
+export class UsbPrinterDeviceChannel extends EventTarget implements IPrinterDeviceChannel {
+    private device: USBDevice;
+    private deviceIn: USBEndpoint;
+    private deviceOut: USBEndpoint;
+
+    public enableConsoleDebug = false;
+
+    private _commMode: PrinterCommMode;
+    public get commMode() {
+        return this._commMode;
+    }
+
+    get channelMode(): PrinterChannelType {
+        return PrinterChannelType.usb;
+    }
+
+    private _ready: Promise<boolean>;
+    public get ready(): Promise<boolean> {
+        return this._ready;
+    }
+
+    private inputStream: ReadableStream<string>;
+    public get streamFromPrinter() {
+        return this.inputStream;
+    }
+
+    public get modelHint(): string {
+        return this.device?.productName;
+    }
+
+    constructor(device: USBDevice, enableConsoleDebug = false) {
+        super();
+
+        this.device = device;
+        this.enableConsoleDebug = enableConsoleDebug;
+
+        this._ready = this.setup();
+    }
+
+    private async setup() {
+        await this.connect();
+        return true;
+    }
+
+    public async dispose() {
+        try {
+            await this.device.close();
+        } catch (e) {
+            if (
+                e instanceof DOMException &&
+                e.name === 'NotFoundError' &&
+                e.message ===
+                    "Failed to execute 'close' on 'USBDevice': The device was disconnected."
+            ) {
+                // Device was already closed, no-op.
+                return;
+            }
+
+            throw e;
+        }
+    }
+
+    public async sendCommands(
+        commandBuffer: Uint8Array
+    ): Promise<PrinterCommunicationError | null> {
+        if (this.enableConsoleDebug) {
+            console.debug('Sending print command buffer to printer via USB..');
+            console.time('sendPrintBuffer');
+        }
+
+        try {
+            // TOOD: Add timeout in case of communication hang.
+            await this.device.transferOut(this.deviceOut.endpointNumber, commandBuffer);
+            return null;
+        } catch (e: unknown) {
+            if (typeof e === 'string') {
```

Comment:
wha?

Reply:
I forget why I wrote this code but when I look at it I get angry, so I'm going to chalk it up to "you can throw anything in JS".